### ISSUE TITLE
0x80 custom authentication handler

### ISSUE THREAD

**Issue by niklas1375** [JoshGlazebrook/socks#65]
I have a very specific SOCKS5 proxy I'd like to connect to. Its authentication method is via JWT as an `0x80` custom authentication method. 
Unfortunately no actively maintained and properly licensed package seems to cover a custom auth handler for `0x80` authentication. I have seen it in some smaller repos that a callback is taken for creating the custom auth call. 
Would it be possible to include such a mechanism in this package?

I'm open for alternative solution ideas as well :)

**Comment by niklas1375** [comment]
Just to clarify: by `0x80` I'm referring to `X'80'` the from the corresponding [RFC](https://www.ietf.org/rfc/rfc1928.txt)

**Comment by JoshGlazebrook** [comment]
I think this should be doable. Essentially would need to extend the options object to allow for a "custom auth type" passed in with values 0x80 - 0xFE. And then a "customAuthHandler" which would be a user-provided function that returns the value (Buffer) to be sent to the proxy server.

Something like
```typescript
const client = await SocksClient.createConnection({
  proxy: {
    host: '1.2.3.4',
    port: 1080,
    type: 5,
  },

  destination: {
    host: '2.3.4.5',
    port: 80,
  },

  command: 'connect',
  custom_auth_method: 0x80,
  custom_auth_handler: async () => {
    const token = '1234abcd'; // = createJWT();
    return Buffer.from(token);
  },
})
```

Would this work?

**Comment by JoshGlazebrook** [comment]
Okay I'll see if I can get this in sometime this week.

**Comment by JoshGlazebrook** [comment]
Okay a few more notes, its become apparent you would also need to have a handler for the server's response.

I also moved the custom_auth_xxxx to the proxy options as this can also be used for creating a chain. 

```typescript
const client = await SocksClient.createConnection({
  proxy: {
    host: '1.2.3.4',
    port: 1080,
    type: 5,
    custom_auth_method: 0x80,
    custom_auth_handler: async () => {
        const token = '1234abcd'; // = createJWT();
        return Buffer.from(token);
    },
    custom_auth_response_size: 2,
    custom_auth_response_handler: async (data) => {
       // data is a <Buffer XX XX> returned from the server after the data from custom_auth_handler() was sent.
       return data[1] === 0x00; // Expects the second byte to be 0x00 if the auth was successful.
    }
  },

  destination: {
    host: '2.3.4.5',
    port: 80,
  },

  command: 'connect',
})
```

Does this seem correct?

**Comment by niklas1375** [comment]
I hadn't thought about the response validation since that's the only "standard" about this proxy. But yes, that would be sufficient. Thank you very much!

**Comment by JoshGlazebrook** [comment]
@niklas1375 Can you check out #66 I've added an example of how this is used. I published to `2.6.0-beta.1` so you can try it out locally. I'll likely merge this in and publish in a couple days.

**Comment by niklas1375** [comment]
Works like a charm :)
Thank you very much!

**Comment by JoshGlazebrook** [comment]
This is now published in `2.6.0`
